client/src/language/middlewareDefinition.ts

```diff
@@ -0,0 +1,147 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) 2023 Savoir-faire Linux. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import { Location, Position, Range, Uri, commands, type LocationLink, type TextDocument } from 'vscode'
+import { type DefinitionMiddleware } from 'vscode-languageclient'
+
+import { getFileContent } from '../lib/src/utils/files'
+import { requestsManager } from './RequestManager'
+import { changeDefinitionUri, checkIsDefinitionRangeEqual, checkIsDefinitionUriEqual, convertToSameDefinitionType, getDefinitionUri, getEmbeddedLanguageDocPosition, getOriginalDocRange } from './utils'
+import { type EmbeddedLanguageDocInfos } from '../lib/src/types/embedded-languages'
+
+export const middlewareProvideDefinition: DefinitionMiddleware['provideDefinition'] = async (document, position, token, next) => {
+  const nextResult = await next(document, position, token)
+  if ((Array.isArray(nextResult) && nextResult.length !== 0) || (!Array.isArray(nextResult) && nextResult !== undefined)) {
+    return nextResult
+  }
+  const embeddedLanguageDocInfos = await requestsManager.getEmbeddedLanguageDocInfos(document.uri.toString(), position)
+  if (embeddedLanguageDocInfos === undefined || embeddedLanguageDocInfos === null) {
+    return
+  }
+  const embeddedLanguageDocContent = await getFileContent(Uri.parse(embeddedLanguageDocInfos.uri).fsPath)
+  if (embeddedLanguageDocContent === undefined) {
+    return
+  }
+  const adjustedPosition = getEmbeddedLanguageDocPosition(
+    document,
+    embeddedLanguageDocContent,
+    embeddedLanguageDocInfos.characterIndexes,
+    position
+  )
+  const vdocUri = Uri.parse(embeddedLanguageDocInfos.uri)
+  const tempResult = await commands.executeCommand<Location[] | LocationLink[]>(
+    'vscode.executeDefinitionProvider',
+    vdocUri,
+    adjustedPosition
+  )
+
+  if (checkIsArrayLocation(tempResult)) {
```

Comment:
Are both branches executing the same code?

Reply:
Yes. They also produce the same compiled javascript code. This check is only to please TypeScript. I'd rather have this pointless check than losing the type assurance provided by TypeScript. That being said, if you look deeper, you'll see all the functions I added in utils.ts begin with `if (definition instanceof Location)`. It shows there could have been two separate branches, but I thought it was easier reason on a single algorithm that does not mind whether the definition is "Location" or "LocationLink".

However, there are still two type assertions in [convertToSameDefinitionType](https://github.com/yoctoproject/vscode-bitbake/pull/38/files#diff-a5ff141411e32f5d167aad66a4678463d637f38307fe7dc22d0ec1cf1ee5109cR128). This is the best I could do.